src/main/java/com/mongodb/spark/sql/connector/read/MongoPartitionReader.java

```diff
@@ -0,0 +1,119 @@
+ * It's responsible for outputting data for a RDD partition.
+ */
+public class MongoPartitionReader implements PartitionReader<InternalRow> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoPartitionReader.class);
+  private final MongoInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoCursor<BsonDocument> mongoCursor;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoPartitionReader(
+      final MongoInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+    this.readConfig = readConfig;
+    LOGGER.debug(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  /** Proceed to next record, returns false if there is no more records. */
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    return getCursor().hasNext();
+  }
+
+  /** Return the current record. This method should return same value until `next` is called. */
+  @Override
+  public InternalRow get() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call get() on a closed PartitionReader.");
+    return bsonDocumentToRowConverter.toInternalRow(getCursor().next());
+  }
+
+  /**
+   * Closes this stream and releases any system resources associated with it. If the stream is
+   * already closed then invoking this method has no effect.
+   */
+  @Override
+  public void close() {
+    if (!closed) {
+      closed = true;
+      releaseCursor();
+    }
+  }
+
+  private MongoCursor<BsonDocument> getCursor() {
+    if (mongoCursor == null) {
+      LOGGER.debug("Opened cursor for partitionId: {}", partition.getPartitionId());
+      mongoClient = readConfig.getMongoClient();
+      mongoCursor =
+          mongoClient
+              .getDatabase(readConfig.getDatabaseName())
+              .getCollection(readConfig.getCollectionName(), BsonDocument.class)
+              .aggregate(partition.getPipeline())
+              .cursor();
+    }
+    return mongoCursor;
+  }
+
+  private void releaseCursor() {
+    if (mongoCursor != null) {
+      LOGGER.debug("Closing cursor for partitionId: {}", partition.getPartitionId());
+      mongoCursor.close();
+      mongoCursor = null;
+      mongoClient.close();
+      mongoClient = null;
```

Comment:
Closing resources this way instead of using the `try`-`finally` or `try`-with-resources statements is prone to resource leaks: if `mongoCursor.close()` completes abruptly, `mongoClient.close()` is not executed.

Reply:
Done

Reply:
Is there a nicer way than the nested try's?

Reply:
Yes, the `try`-with-resources statement can execute a sequence of actions expressed as `AutoCloseable` objects in a way similar to nested `try`-`finally` statements without nesting.

P.S. Irrelevant for our case, and relevant only in situations when the number of actions is not known at compile time: Guava has [`Closer](https://guava.dev/releases/19.0/api/docs/com/google/common/io/Closer.html) that may execute a dynamic number of actions with behavior similar to that of the `try`-with-resources statement.